simulatingrisk/hawkdove/model.py

```diff
@@ -0,0 +1,375 @@
+divergent_colors_5 = ["#d7191c", "#fdae61", "#ffffbf", "#a6d96a", "#1a9641"]
+
+
+class HawkDoveAgent(mesa.Agent):
+    """
+    An agent with a risk attitude playing Hawk or Dove
+    """
+
+    def __init__(self, unique_id, model, hawk_odds=None):
+        super().__init__(unique_id, model)
+
+        self.points = 0
+        self.choice = self.initial_choice(hawk_odds)
+        self.last_choice = None
+
+        # risk level must be set by base class, since initial
+        # conditions are specific to single / variable risk games
+        self.set_risk_level()
+
+    def set_risk_level(self):
+        raise NotImplementedError
+
+    def __repr__(self):
+        return (
+            f"<{self.__class__.__name__} id={self.unique_id} "
+            + f"r={self.risk_level} points={self.points}>"
+        )
+
+    def initial_choice(self, hawk_odds=None):
+        # first round : choose what to play randomly or based on initial hawk odds
+        opts = {}
+        if hawk_odds is not None:
+            opts["weight"] = hawk_odds
+        return coinflip(play_choices, **opts)
+
+    @property
+    def choice_label(self):
+        return "hawk" if self.choice == Play.HAWK else "dove"
+
+    def get_neighbors(self, size):
+        """get all neighbors for a supported neighborhood size"""
+        check_neighborhood_size(size)
+        # 4 and 8 neighborhood use default radius 1
+        # 8 and 24 both use moore neighborhood (includes diagonals)
+        opts = {"moore": True}
+        if size == 4:
+            # use von neumann neighborhood instead of moore (no diagonal)
+            opts["moore"] = False
+
+        # for 24 size neighborhood, use radius 2
+        if size == 24:
+            opts["radius"] = 2
+
+        return self.model.grid.get_neighbors(self.pos, include_center=False, **opts)
+
+    @property
+    def play_neighbors(self):
+        """neighbors to play against, based on model play neighborhood size"""
+        return self.get_neighbors(self.model.play_neighborhood)
+
+    @property
+    def observed_neighbors(self):
+        """neighbors to look at when deciding what to play;
+        based on model observed neighborhood size"""
+        return self.get_neighbors(self.model.observed_neighborhood)
+
+    @property
+    def num_dove_neighbors(self):
+        """count how many neighbors played DOVE on the last round
+        (uses `observed_neighborhood` size from model)"""
+        return len([n for n in self.observed_neighbors if n.last_choice == Play.DOVE])
+
+    @property
+    def proportional_num_dove_neighbors(self):
+        """adjust the number of dove neighbors based on ratio between
+        play neighborhood and observed neighborhood, to scale observations
+        to the range of agent risk level."""
+        ratio = self.model.max_risk_level / self.model.observed_neighborhood
+        # always round to an integer
+        return round(ratio * self.num_dove_neighbors)
```

Comment:
I think the general strategy of scaling the # of doves to the range of risk levels which corresponds to the "base" neighborhood of 8 is perfectly fine, though it may benefit from being noted down in narrative docs/READMEs in order to explain how a choice is made when observed neighborhood is greater or less than 8